engine/consensus/matching/engine.go

```diff
@@ -1225,9 +1317,35 @@ func (e *Engine) requestPendingApprovals() (int, error) {
 		}
 	}
 
-	if requestCount > 0 {
-		log.Debug().Msgf("requested %d approvals", requestCount)
+	return requestCount, nil
+}
+
+// OnFinalizedBlock implements the callback from the protocol state to notify a block
+// is finalized, it guarantees every finalized block will be called at least once.
+func (e *Engine) OnFinalizedBlock(block *model.Block) {
+	// we index the execution receipts by the executed block ID only for all finalized blocks
+	// that guarantees if we could retrieve the receipt by the index, then the receipts
+	// must be for a finalized blocks.
+	err := e.indexReceipts(block.BlockID)
+	if err != nil {
+		e.log.Fatal().Err(err).Hex("block_id", block.BlockID[:]).
+			Msg("could not index receipts for block")
+	}
+}
+
+func (e *Engine) indexReceipts(blockID flow.Identifier) error {
```

Comment:
do we need to do this as part of matching engine? Seems a little bit out of place.

Reply:
Yes, otherwise the request pending receipts won't work well. 

[See details here(https://github.com/onflow/flow-go/pull/407/files#diff-17e93dc346462f66b95ec3f31134583765b8fdc4eff5e3c03f276698a554997fR1124-R1128)